A4A > Referrals: Implement referral details view

yashwin opened the pull request:
Closes https://github.com/Automattic/jetpack-genesis/issues/347

## Proposed Changes

This PR implements UI for the referral details view

NOTE: 

- [Design](https://www.figma.com/design/fuufP6VNfZXYmvLsTRWRlG/Referrals?node-id=6711-71510&m=dev)
- There are some UI enhancement & mobile fixes that will be done in another PR, so please ignore those for now.
## Testing Instructions

1. Open the A4A live link.
2. Append the /referrals URL as /referrals?flags=-a4a-automated-referrals, and verify that the above changes are not reflected and that the Referrals page matches the production.
3. Remove the feature flag > Go to  Referrals - Dashboard.
4. Use the below API endpoint manually to create a referral if you don't have some. Create a couple of them

POST https://public-api.wordpress.com/wpcom/v2/agency/230591090/referrals?client_email={client_email}&client_message={your_message}&product_ids={product_id,products_2}

5. Patch D149816
6. Now visit the  Referrals - Dashboard & refresh the page > Once you see the list of referrals, click the `>` icon & verify the page looks like below. It will be mostly empty in this PR.

<img width="1647" alt="Screenshot 2024-05-27 at 2 15 46 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/8bbfc598-b31d-4f38-983a-5bbb9a236cdc">


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

yashwin committed:
Implement Referral Details view

cleacos commented on client/a8c-for-agencies/components/items-dashboard/item-preview-pane/item-preview-pane-header/index.tsx:
```diff
@@ -44,32 +46,38 @@ export default function ItemPreviewPaneHeader( {
-					className="item-preview__header-favicon"
-					size={ size }
-				/>
+				{ isSiteData && (
+					<SiteFavicon
+						blogId={ itemData.blogId }
+						fallback={ siteIconFallback }
+						color={ itemData.color }
+						className="item-preview__header-favicon"
+						size={ size }
+					/>
+				) }
```
I know it still has some linked pieces for specific contexts, but if we can keep this component totally agnostic, better.

So instead, I'd introduce this property in the `ItemData` object: `withIcon` (by default true), and if it's false, we won't render the icon.

cleacos commented on client/a8c-for-agencies/components/items-dashboard/item-preview-pane/item-preview-pane-header/index.tsx:
```diff
@@ -44,32 +46,38 @@ export default function ItemPreviewPaneHeader( {
-									/>
-								</span>
-							</Button>
+							{ isSiteData ? (
+								<Button
+									variant="link"
+									className="item-preview__header-summary-link"
+									href={ itemData.url }
+									target="_blank"
+								>
+									<span>
+										{ itemData.subtitle }
+										<Icon
+											className="sidebar-v2__external-icon"
+											icon={ external }
+											size={ extraProps?.externalIconSize || ICON_SIZE_SMALL }
+										/>
+									</span>
+								</Button>
+							) : (
+								itemData.subtitle
+							) }
```
And here is something similar to the above suggestion, `subtitleWithLink` or even better, just check if `itemData.url`  is not empty/undefined to determine to use the Button+Icon or just render the subtitle.

cleacos commented on client/a8c-for-agencies/components/items-dashboard/item-preview-pane/types.ts:
```diff
@@ -19,13 +19,14 @@ export interface FeatureTabInterface {
 	blogId?: number;
 	isDotcomSite?: boolean;
 	adminUrl?: string;
+	isSiteData?: boolean;
```
`withIcon`

cleacos commented on client/a8c-for-agencies/sections/referrals/referral-details/index.tsx:
```diff
@@ -0,0 +1,81 @@
+export default function ReferralDetails( { referral, closeSitePreviewPane }: Props ) {
+	const translate = useTranslate();
+
+	const [ selectedSiteFeature, setSelectedSiteFeature ] = useState( REFERRAL_PURCHASES_ID );
```
Rename suggestion: `selectedReferralTab` and `setSelectedReferralTab`

cleacos commented on client/a8c-for-agencies/sections/referrals/referrals-list/index.tsx:
```diff
@@ -92,7 +124,9 @@ export default function ReferralList( { referrals, dataViewsState, setDataViewsS
 					totalPages: 1,
 				},
-				itemFieldId: 'ref',
+				getItemId: ( item: any ) => {
+					return item.id;
+				},
```
If I recall well, if the item has the `id` field, that works perfectly with the DataViews component, if not, we have two alternatives:

- `itemFieldId: 'ref'` even it supports nested fields, `itemFieldId: 'site.value.blog_id'`
- Using the `getItemId` callback method.

So in theory, `getItemId` is not required because the item has the `id` field.

cleacos reviewed (changes requested):
Good job!

It works as expected, but... the scroll doesn't work on the Referral page when you deactivate the feature flag (current version).

Also, I left a few comments/recomendations related to the `ItemsDataViews` and `ItemPreviewPane` usage.

yashwin committed:
Address PR comments

yashwin commented:
Thanks for the thorough review, @cleacos!

> It works as expected, but... the scroll doesn't work on the Referral page when you deactivate the feature flag (current version).

This is fixed now: https://github.com/Automattic/wp-calypso/pull/91157/commits/a18663ce33fa8f5feff69c1ee78b6a8569fc8ca9

> Also, I left a few comments/recomendations related to the `ItemsDataViews` and `ItemPreviewPane` usage.

I have addressed all your comments: https://github.com/Automattic/wp-calypso/pull/91157/commits/a18663ce33fa8f5feff69c1ee78b6a8569fc8ca9

cleacos reviewed (approved):
Cool! Now, the scroll works as expected. The rest continues working well after the changes.

LGTM!
